Completed Ubuntu Headless Setup with QT Library Integration

long-trinh-rvc opened the pull request:
This pull request finishes the setup for an Ubuntu headless environment, with the QT library successfully copied into the image.

vudangRVC commented on ubuntu-headless-qt/README.md:
The readme should show an overview of folder structure and the purpose of the build folder.
You can refer to: [README.md](https://github.com/Renesas-SST/rz-build-scripts/blob/dunfell/rz-sbc/README.md)

is ubuntu 22.04 LXDE correct?

long-trinh-rvc replied:
I will change the title to "ubuntu 22.04 headless with Qt library" and add folder structure to README.md

vudangRVC replied:
Can you rename "config.sh" to "config.ini"?
This is an example of how to use config.ini: https://github.com/seebe/rzg_stuff/blob/master/build_scripts/multimedia-modules/config.ini

vudangRVC replied:
How can I perform a build? I don't see any mention of that.

vudangRVC replied:
@long-trinh-rvc ,
Also, we need an introduction section for the image. Refer to QTv1.1.1 package or Ubuntu development guide from the reference board are ok.

long-trinh-rvc replied:
The build instruction is being added in the next commit. I have also changed the name to config.ini

vudangRVC replied:
Don't copy the whole readme from QTv1.1.1 package which is intended to provide the environment on Yocto Poky Distro. This will confuse the user that there is any difference between QTv1.1.1 package and this package?

Make sure you understand the Ubuntu porting procedure, then update the guideline like a porting procedure.
You can refer to "Remi Pi_Ubuntu System Development Guide" to know how they introduce their porting process.

The features section is ok.

vudangRVC commented on ubuntu-headless-qt/include/00_prepare_env.sh:
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
+
```
Please add a description for this script.

vudangRVC commented on ubuntu-headless-qt/README.md:
```diff
@@ -0,0 +1,5 @@
+A core-image-qt-rzpi.tar.bz2 file needed for build script
+Put that file inside ubuntu-headless-qt folder
+
+Please remove rootfs and rootfs_qt before running main_script.sh
```
I'm not sure what are rootfs and rootfs_qt?
As my understanding, there are two rootfs. One is the target rootfs and other is original qt rootfs. Is it correct?
If so, please change the folder name to make more sense.

long-trinh-rvc replied:
In this current script:
rootfs (target)
rootfs_qt (source)

I will change to:
rootfs (target)
qt_rootfs_source (source)
to be easier to make more sense

long-trinh-rvc replied:
And I will also remove 
~~Please remove rootfs and rootfs_qt before running main_script.sh~~
Because there is a function that handle this feature

vudangRVC commented on ubuntu-headless-qt/include/01_prepare_ubuntu_base.sh:
```diff
@@ -0,0 +1,136 @@
+#!/bin/bash
+# --------------------------------------------------------------------------#
+# function ubuntu_base_prepare use to prepare basic binaries of ubuntu os.
```
This description looks like a function description. Can you update it and move the old description to the correct location of the mentioned functions?

vudangRVC commented on ubuntu-headless-qt/include/02_prepare_rootfs_qt.sh:
```diff
@@ -0,0 +1,208 @@
+#!/bin/bash
+# --------------------------------------------------------------------------#
+# function rootfs_qt use to copy binaries of rootfs_qt to ubuntu os.
```
This description looks like a function description. Can you update it and move the old description to the correct location of the mentioned functions?

vudangRVC commented on ubuntu-headless-qt/include/03_prepare_conf.sh:
```diff
@@ -0,0 +1,168 @@
+#!/bin/bash
+# --------------------------------------------------------------------------#
+# function set_config use to copy config file to ubuntu os.
```
This description looks like a function description. Can you update it and move the old description to the correct location of the mentioned functions?

vudangRVC commented on ubuntu-headless-qt/include/04_mount.sh:
```diff
@@ -0,0 +1,154 @@
+#!/bin/bash
+# --------------------------------------------------------------------------#
+# function run_in_chroot use to run script in ubuntu os after mount chroot.
```
This description looks like a function description. Can you update it and move the old description to the correct location of the mentioned functions?

vudangRVC commented on ubuntu-headless-qt/include/05_create_wic.sh:
```diff
@@ -0,0 +1,89 @@
+#!/bin/bash
+# --------------------------------------------------------------------------#
+# function create_wic use to create wic from ubuntu rootfs.
```
This description looks like a function description. Can you update it and move the old description to the correct location of the mentioned functions?

vudangRVC commented on ubuntu-headless-qt/main_script.sh:
```diff
@@ -0,0 +1,72 @@
+#!/bin/bash
+# --------------------------------------------------------------------------------#
+# Main function 
```
Add script description

vudangRVC commented on ubuntu-headless-qt/main_script.sh:
```diff
@@ -0,0 +1,72 @@
+source include/05_create_wic.sh
+
+# main function
+function main(){
```
Add comment for each function call

vudangRVC commented on ubuntu-headless-qt/script/apt_install_base.sh:
```diff
@@ -0,0 +1,57 @@
+#!/bin/bash
+# Check if the script is run as root
```
Add script description

vudangRVC commented on ubuntu-headless-qt/script/apt_wifi_ble.sh:
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+# Check if the script is run as root
```
Add script description

long-trinh-rvc replied:
This script is redundant and will be removed

vudangRVC commented on ubuntu-headless-qt/script/create_rzpi_user.sh:
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
+
```
Add script description

vudangRVC commented on ubuntu-headless-qt/script/set_root_password.sh:
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
+
```
Add script description

sonnguyenxg replied:
Hello @vudangRVC , @long-trinh-rvc,
I updated.

vudangRVC commented on ubuntu-headless-qt/script/set_root_password.sh:
```diff
@@ -0,0 +1,31 @@
+    echo "Please run this script as root or using sudo."
+    exit 1
+fi
+
```
Add explanation for these steps

sonnguyenxg replied:
Hello @vudangRVC , @long-trinh-rvc,
I updated.

vudangRVC commented on ubuntu-headless-qt/script/create_rzpi_user.sh:
```diff
@@ -0,0 +1,34 @@
+    exit 1
+fi
+
+# Variables
```
Should we move these variables into a "configuration" file? So Users can set them to whatever they want instead of hardcoding.

long-trinh-rvc replied:
Users can create themselves a new user so I will remove this feature

sonnguyenxg replied:
Hello @vudangRVC , @long-trinh-rvc ,
The root user can't run the lxde-desktop. so we need to create a new normal user.

vudangRVC commented on ubuntu-headless-qt/script/apt_wifi_ble.sh:
```diff
@@ -0,0 +1,24 @@
+
+# time zone data
+echo 'tzdata tzdata/Areas select Asia' | sudo debconf-set-selections
+echo 'tzdata tzdata/Zones/Asia select Ho_Chi_Minh' | sudo debconf-set-selections
```
Should we move these variables into a "configuration" file? So Users can set them to whatever they want instead of hardcoding.

long-trinh-rvc replied:
This script is redundant and will be removed

vudangRVC commented on ubuntu-headless-qt/include/05_create_wic.sh:
```diff
@@ -0,0 +1,89 @@
+    OUTPUT_WIC="ubuntu-image-qt-rzpi.wic"
+
+    # Define Size
+    BOOT_SIZE_MB=200
```
Should we move these variables into a "configuration" file? So Users can set them to whatever they want instead of hardcoding.

Variables:
- OUTPUT_WIC
- BOOT_SIZE_MB

vudangRVC replied:
The configuration file could be "config.ini" and it should be located at the same level as the main script

long-trinh-rvc replied:
About the configuration file, I will use config.sh so main script can `source config.sh` and get those variables easier

vudangRVC commented on ubuntu-headless-qt/include/03_prepare_conf.sh:
```diff
@@ -0,0 +1,168 @@
+    cd "$WORK_DIR" || { echo "Failed to change to WORK_DIR"; return 1; }
+
+    # Call Function
+    copy_qemu
```
Add comment for each function call

vudangRVC commented on ubuntu-headless-qt/include/02_prepare_rootfs_qt.sh:
```diff
@@ -0,0 +1,208 @@
+    echo "tar_core_image_qt completed successfully."
+
+    echo "5. Starting copy_boot_folder and copy_qt, copy_wifi_firmware..."
+    copy_boot_folder
```
Add comment for each function call to introduce the purpose shortly

vudangRVC commented on ubuntu-headless-qt/include/01_prepare_ubuntu_base.sh:
```diff
@@ -0,0 +1,136 @@
+    fi
+
+    # Check and download file ubuntu-base
+    local file_name="ubuntu-base-22.04-base-arm64.tar.gz"
```
These variables should be moved to "config.ini" file

long-trinh-rvc commented:
Hi @vudangRVC ,
I have updated based on your comments, please help me review latest commit.

vudangRVC commented on ubuntu-headless-qt/script/apt_install_base.sh:
```diff
@@ -0,0 +1,74 @@
+TIME_ZONE_AREA="${TIME_ZONE_AREA:=tzdata tzdata/Areas select Asia}"
+TIME_ZONE_CITY="${TIME_ZONE_CITY:=tzdata tzdata/Zones/Asia select Ho_Chi_Minh}"
+echo 'tzdata tzdata/Areas select Asia' | sudo debconf-set-selections
+echo 'tzdata tzdata/Zones/Asia select Ho_Chi_Minh' | sudo debconf-set-selections
```
These two echo commands are hardcoded, correct?

long-trinh-rvc replied:
This is my mistake, I have fix that in the latest commit.

vudangRVC replied:
```
echo 'tzdata tzdata/Areas select Asia' | sudo debconf-set-selections
echo 'tzdata tzdata/Zones/Asia select Ho_Chi_Minh' | sudo debconf-set-selections
```

Can you explain these commands? Do we need to set 'tzdata tzdata/Areas select Asia' and 'tzdata tzdata/Zones/Asia select Ho_Chi_Minh'? I think they are already set in config.ini

long-trinh-rvc replied:
> ```
> echo 'tzdata tzdata/Areas select Asia' | sudo debconf-set-selections
> echo 'tzdata tzdata/Zones/Asia select Ho_Chi_Minh' | sudo debconf-set-selections
> ```
> 
> Can you explain these commands? Do we need to set 'tzdata tzdata/Areas select Asia' and 'tzdata tzdata/Zones/Asia select Ho_Chi_Minh'? I think they are already set in config.ini

Those command are redundant. I have changed to use time zone parameters set in config.ini

vudangRVC replied:
`tzdata tzdata/Zones/Asia`

Is this "Asia" a hardcode?

long-trinh-rvc replied:
> `tzdata tzdata/Zones/Asia`
> 
> Is this "Asia" a hardcode?

It's my mistake again. In next commit those settings will be changed to:
```c
echo "tzdata tzdata/Areas select $TIME_ZONE_AREA" | sudo debconf-set-selections
echo "tzdata tzdata/Zones/$TIME_ZONE_AREA select $TIME_ZONE_CITY" | sudo debconf-set-selections
```

long-trinh-rvc commented:
Hi @vudangRVC ,
I have uploaded new commit based on your comments, please help me review it.

vudangRVC commented on ubuntu-headless-qt/README.md:
```diff
@@ -0,0 +1,996 @@
+# RZG2L SBC board #
```
Update this section to show that we are porting and creating a custom Ubuntu-core image for RZG2L-SBC

vudangRVC replied:
The first line "RZG2L SBC board" should change to "Ubuntu Core System development on RZG2L SBC board"

vudangRVC commented on ubuntu-headless-qt/README.md:
```diff
@@ -0,0 +1,996 @@
+The below will describe the current status of development, how to build, set up environment for RZG2L-SBC.
+
+## Status
+This is a Ubuntu release of the RZG2L development product for RZG2L-SBC.
```
It's a Custom Ubuntu-core release

vudangRVC commented on ubuntu-headless-qt/README.md:
```diff
@@ -0,0 +1,996 @@
+
+This release provides the following features:
+
+ - Ubuntu build compatible with RZG2L SoC
```
What is "Ubuntu build compatible with RZG2L SoC"?

In case of Yocto project, we have all configurations for RZG2L SoC. But in this case, what is the build system that you are using? You just create build scripts and that's all?

long-trinh-rvc replied:
I will change to :
```
 - Custom Ubuntu Core build scripts for easy setup and deployment.
```

vudangRVC commented on ubuntu-headless-qt/README.md:
```diff
@@ -0,0 +1,996 @@
+
+ - Only support for 48 Khz audio sampling rate family.
+
+## Building Ubuntu
```
Are these steps for building Yocto? Why are you putting them here?

If your expectation is just let users know that the "core-image-qt-rzpi.tar.bz2" is required for the build. Just mention that.

long-trinh-rvc replied:
These steps are for running main_script.sh, I will remove some redundant guides (such as install package and setup git configuration)

long-trinh-rvc replied:
I have replaced all step below to the new one, referring Ubuntu System Porting Guide for Renesas RZ/G2L-based Remi Pi

vudangRVC commented on ubuntu-headless-qt/README.md:
````diff
@@ -0,0 +1,996 @@
+|-- ubuntu-image-qt-rzpi.wic                <---- Output WIC
+`-- ubuntu-image-qt-rzpi.wic.tar.gz         <---- Output compressed WIC
+```
+### U-boot environment
````
How can I prepare U-Boot environment?

You can ask users refer to the original package as all images have same the procudure

vudangRVC replied:
Please point out the link for "the original packages".

long-trinh-rvc replied:
Do you mean the link to Readme in QTv1.1.1 package ?

vudangRVC replied:
I accept this change for now. I'm not sure how we organize the release package. Let's discuss with US team on this

vudangRVC commented on ubuntu-headless-qt/README.md:
````diff
@@ -0,0 +1,996 @@
+To set GPIO pin, move to GPIO sysfs directory and set values as shown below:
+
+```
+root@rzpi:~# cd /sys/class/gpio/
````
I saw from your verification logs. Sometimes, the prompt is 
`root@localhost:~#`
or 
`rzsbc@localhost:`

There is no
`root@rzpi:~#`

Which one is correct?

long-trinh-rvc replied:
`rzsbc@localhost` is the correct one, I will change from `root@rzpi` to `rzsbc@localhost`

vudangRVC commented on ubuntu-headless-qt/README.md:
````diff
@@ -0,0 +1,996 @@
+root@rzpi:~# apt-get install -f
+```
+
+### Network Boot and TFTP
````
Remove this section; I don't think Ubuntu users will find this feature interesting.

vudangRVC replied:
Also, can you check if any majority feature on Ubuntu we can show to impress users?

Check "minimal ubuntu" keyword

long-trinh-rvc replied:
I will add netplan configuration in new ubuntu release

vudangRVC commented on ubuntu-headless-qt/README.md:
````diff
@@ -0,0 +1,996 @@
+  Starting kernel ...
+  ```
+
+### Using SSH and SCP for Remote Access and File Transfers
````
Remove this section; I don't think Ubuntu users will find this feature interesting.

long-trinh-rvc commented:
Hi @vudangRVC ,
I have updated README and fixed some hardcode mistake. Please review my latest update.

vudangRVC commented on ubuntu-headless-qt/README.md:
```diff
@@ -0,0 +1,681 @@
+# RZG2L SBC board #
+This is the quick startup guide for RZG2L SBC board (hereinafter referred to as `RZG2L-SBC`).
```
"This is the quick startup guide for RZG2L SBC board (hereinafter referred to as `RZG2L-SBC`)."

Change to something like

"This is the quick startup guide for RZG2L SBC board (hereinafter referred to as `RZG2L-SBC`) to develop on Ubuntu headless (no Desktop environment support)"

vudangRVC commented on ubuntu-headless-qt/README.md:
```diff
@@ -0,0 +1,681 @@
+### Introduction
+Ubuntu-base is the minimum file system officially built by Ubuntu, which includes the Debian package manager. The size of the base package is usually only tens of megabytes, behind which there is the entire ubuntu software repository support. Ubuntu software generally has good stability. Based on Ubuntu-base, Linux software can be installed on demand, with deep customization capabilities, and it  ...
+
+
```
Remove this empty line

vudangRVC commented on ubuntu-headless-qt/README.md:
```diff
@@ -0,0 +1,681 @@
+
+Several common methods for building embedded file systems include busybox, yocto and buildroot. But Ubuntu offers a convenient and powerful package management system with strong community support, allowing for the installation of new software packages directly through apt-get install. This article describes how to build a complete Ubuntu system based on Ubuntu-base. Ubuntu supports many architect ...
+
+Before starting the porting, prepare a file named `core-image-qt-rzpi.tar.bz2`
```
Where can I get this file? You should indicate a specific example.

vudangRVC commented on ubuntu-headless-qt/README.md:
````diff
@@ -0,0 +1,681 @@
+
+Then we can execute the script as follows:
+```
+chmod +x main_script.sh
````
I forgot that we need to point out that we have a configuration file. 
Please add some description for that. And how users can use the file.

vudangRVC commented on ubuntu-headless-qt/README.md:
````diff
@@ -0,0 +1,681 @@
+Then we can execute the script as follows:
+```
+chmod +x main_script.sh
+./main_script.sh
````
Can you create a table summarizing which packages are installed while the script runs?

No need for license description but the purpose of the packages

long-trinh-rvc commented:
Hi @vudangRVC ,
I have updated based on your comments, please review my latest commit.

vudangRVC commented:
@long-trinh-rvc ,

I didn't notice that you installed Gstreamer from apt. Please ensure that GPU, Video codec library, Weston, and Gstreamer are ported to the rootfs properly.

Also, add an option in config.ini to enable Weston desktop (disabled by default as we create a headless image)

And, the script needs to be updated to support Yocto build as discussed.

vudangRVC commented on ubuntu-headless-qt/README.md:
````diff
@@ -0,0 +1,702 @@
+Run the following commands to play an audio file:
+
+```
+root@localhost:~# aplay /home/root/audios/04_16KH_2ch_bgm_maoudamashii_healing01.wav
````
@long-trinh-rvc ,

We don't have these audio files anymore. Please update.

long-trinh-rvc replied:
I've updated it in the latest commit.

tiennguyenuh commented on ubuntu-headless-qt/include/common/yocto_working.sh:
```diff
@@ -0,0 +1,62 @@
+# This function help main script can build yocto
+build_yocto() {
+    # Get source yocto
+    su -c "bash -c 'source include/common/yocto_source.sh; mkdir yocto_rzsbc_board ; cd yocto_rzsbc_board ; get_bsp'" "$MAIN_USER"
```
Please remove the su command, as it is not suitable when using sudo privileges during Yocto builds.

long-trinh-rvc replied:
Without the su command, the script cannot be executed to build Yocto due to 'root privileges' when we run 
`sudo ./main_script.sh`
This command keeps yocto build running under "$MAIN_USER", not root (sudo)

tiennguyenuh replied:
With this command, we cannot build in the CloudBees environments because the normal user in CloudBees does not have root privileges. Please try a different approach to handle that case.

tiennguyenuh commented on ubuntu-headless-qt/include/common/yocto_working.sh:
```diff
@@ -0,0 +1,62 @@
+    # Loop until we find the output file
+    while [ -z "$result" ]; do
+        # Run bitbake
+        su -c "bash -c 'source include/common/yocto_source.sh; cd yocto_rzsbc_board ; setup_conf; \
```
Please remove the su command, as it is not suitable when using sudo privileges during Yocto builds.

tiennguyenuh commented on ubuntu-headless-qt/main_script.sh:
```diff
@@ -0,0 +1,312 @@
+
+# main function for ubuntu core
+main_ubuntu_core(){
+    MAIN_USER=$(sudo grep 'sudo: .*main_script.sh' /var/log/auth.log | tail -n 1 | awk '{print $6}')
```
Yocto part should be moved out of this function, encapsulated in a separate function, and then called from main.

tiennguyenuh commented on ubuntu-headless-qt/main_script.sh:
```diff
@@ -0,0 +1,312 @@
+#   None
+#######################################
+main_ubuntu_lxde(){
+    MAIN_USER=$(sudo grep 'sudo: .*main_script.sh' /var/log/auth.log | tail -n 1 | awk '{print $6}')
+    # Recheck user for yocto build
+    if [ -n "$MAIN_USER" ]; then
+        echo "User executed sudo ./main_script is: $MAIN_USER"
+    else
+        echo "It seem that you are root. Recheck..."
+        MAIN_USER=$(stat -c '%U' main_script.sh)
+        if [ -n "$MAIN_USER" ]; then
+            echo "User executed sudo ./main_script is: $MAIN_USER"
+        else
+            echo "It seem that you are root. Please login and clone as a user"
+            exit 1
+        fi
+    fi
+
+    if [ "$MAIN_USER" = "root" ]; then
+        echo "Error: Current user cannot be root, we cannot build yocto with root's privilege."
+        exit 1
+    fi
+
+    ######## YOCTO WORKING ########
+    build_yocto
+    if [ $? -eq 1 ]; then
+        echo "build_yocto failed."
+        exit 1
+    fi
+    ##### END YOCTO WORKING ######
```
Same, Yocto part should be moved out of this function, encapsulated in a separate function, and then called from main.

tiennguyenuh commented on ubuntu-headless-qt/config.ini:
```diff
@@ -0,0 +1,51 @@
+# UBUNTU_BASE_LINK="https://cdimage.ubuntu.com/ubuntu-base/releases/24.04/release/$UBUNTU_BASE_FILE_NAME"
+
+# Define output rootfs file name
+OUTPUT_ROOTFS="rootfs"
+
+# Define output wic file name
+OUTPUT_WIC="ubuntu-image-qt-rzpi.wic"
```
We should use the same name for the output file, for example:
- OUTPUT_ROOTFS: ubuntu-image-qt-rzpi.tar.bz2
- OUTPUT_WIC: ubuntu-image-qt-rzpi.wic

`ubuntu-image-qt-rzpi` should be defined in another variable

vudangRVC commented:
@long-trinh-rvc ,

After complete the updates, please squash these commits into 4 separate commits:
1. Common scripts
2. Updates for Ubuntu-core
3. Updates for Ubuntu-lxde
4. Yocto/Cloudbee build updates

long-trinh-rvc committed:
rz-sbc: Add common base script for Ubuntu build script

    This commit introduces a common base script that streamlines the
    process of building Ubuntu images for the rz-sbc platform.

long-trinh-rvc committed:
rz-sbc: Add Ubuntu Core build script

    This commit adds a new build script specifically designed for
    creating Ubuntu Core images for the rz-sbc platform.

long-trinh-rvc committed:
rz-sbc: Add Ubuntu LXDE build script

    This commit adds a new build script specifically designed for
    creating Ubuntu LXDE images for the rz-sbc platform.

long-trinh-rvc committed:
rz-sbc: Add Yocto build to support Ubuntu build script, ensure compatibility for CloudBees environment

    This commit integrates Yocto build capabilities into the existing
    Ubuntu build script for the rz-sbc platform.

tiennguyenuh commented on ubuntu/include/common/yocto_working.sh:
Should we use the yocto source script from [rzsbc_yocto.sh](https://github.com/Renesas-SST/rz-build-scripts/blob/dunfell/rz-sbc/rz-sbc/rzsbc_yocto.sh) instead?

long-trinh-rvc replied:
The script now uses [rzsbc_yocto.sh](https://github.com/Renesas-SST/rz-build-scripts/blob/dunfell/rz-sbc/rz-sbc/rzsbc_yocto.sh) by using sed function to clone a new one to Ubuntu script

tiennguyenuh committed:
rz-sbc: use meta-renesas-sst for development purpose

Signed-off-by: Tien Nguyen <tien.nguyen.uh@renesas.com>

tiennguyenuh committed:
rz-sbc: change zst to bz2 type of root filesystem

Signed-off-by: Tien Nguyen <tien.nguyen.uh@renesas.com>

long-trinh-rvc committed:
rz-sbc: Add lxde apt install script, note that this branch will not be used anymore after 3 PRs on SST repo are merged

long-trinh-rvc committed:
ubuntu : fix bug at LXDE build
